### PR TITLE
ROX-20788: add wif client to google registry

### PR DESCRIPTION
**Pull request by stehessel**
## Description

The goal of this PR is to enable workload identities for Google registries. The workload identity tokens are managed globally by the GCP token manager.

The PR contains the following changes:

* A new option `TokenSource` is added to the config of docker registries. If it is set, the docker transport will be wrapped with a oauth2 transport based on the token source.
* For Google registries, we now use the token source for authentication. This means that for static tokens, an access token is used instead of sending the raw service account JSON. For WIF registries we use the token manager as a token source.
* Much of the remaining diff is caused by bubbling up `gcp.Singleton` to `pkg/registries` from both Central and Sensor.

Note that dedicated end-to-end tests for WIF integrations will be added as part of https://issues.redhat.com/browse/ROX-21034.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

### Here I tell how I validated my change

Tested on a test OpenShift cluster (ROSA HCP):
* wif enabled, initial token and token refresh after 1 hour
* static credentials, initial token and token refresh after 1 hour

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

**Comment by stehessel** [comment]
Current dependencies on/for this PR:
* master
  * **PR #9002** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/9002?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #8949** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/8949?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/8949?utm_source=stack-comment).

File: pkg/registries/docker/docker.go [hunk 1]
```diff
@@ -105,6 +112,9 @@ func NewDockerRegistryWithConfig(cfg Config, integration *storage.ImageIntegrati
 	} else {
 		transport = proxy.RoundTripper()
 	}
+	if cfg.TokenSource != nil {
```
Comment (stehessel) [inline pkg/registries/docker/docker.go:115]: This is where the token source added to the docker registry transport.

File: pkg/registries/google/google.go [hunk 2]
```diff
@@ -32,50 +37,82 @@ func (g *googleRegistry) Match(image *storage.ImageName) bool {
...
-func Creator() (string, func(integration *storage.ImageIntegration) (types.Registry, error)) {
-	return "google", func(integration *storage.ImageIntegration) (types.Registry, error) {
-		return NewRegistry(integration, false)
-	}
+func Creator() (string,
+	func(integration *storage.ImageIntegration, options *types.CreatorOptions) (types.Registry, error),
+) {
+	return "google",
+		func(integration *storage.ImageIntegration, options *types.CreatorOptions) (types.Registry, error) {
+			return NewRegistry(integration, false, options.GetGCPTokenManager())
```
Comment (stehessel) [inline pkg/registries/google/google.go:45]: This is where we pipe in the gcp token manager.

File: pkg/registries/artifactory/artifactory.go [hunk 3]
```diff
@@ -7,16 +7,22 @@ import (
 )
 
 // Creator provides the type and registries.Creator to add to the registry of image registries.
-func Creator() (string, func(integration *storage.ImageIntegration) (types.Registry, error)) {
-	return "artifactory", func(integration *storage.ImageIntegration) (types.Registry, error) {
-		return docker.NewDockerRegistry(integration, false)
-	}
+func Creator() (string,
+	func(integration *storage.ImageIntegration, _ *types.CreatorOptions) (types.Registry, error),
```
Comment (dhaus67) [inline pkg/registries/artifactory/artifactory.go:11]: If we touch this anyways, it'd be nice to make this function an alias in `types`.

File: pkg/registries/types/types.go [hunk 4]
```diff
@@ -57,3 +58,17 @@ var DockerfileInstructionSet = map[string]struct{}{
 	"VOLUME":      {},
 	"WORKDIR":     {},
 }
+
+// CreatorOptions specifies optional configuration parameters for a registry creators.
+type CreatorOptions struct {
```
Comment (dhaus67) [inline pkg/registries/types/types.go:63]: I think the options struct is fine, however I'd rather pass in optional functions in the creator function instead of a required type.

One good example of this would be [the provider options](https://github.com/stackrox/stackrox/blob/master/pkg/auth/authproviders/provider_options.go) for the auth provider creation: if you implement it this way, existing usages will stay as-is and not require changes, and only places where the option is required to be set will have changes.

**Review by dhaus67 — CHANGES_REQUESTED** [review]

File: pkg/registries/artifactory/artifactory.go [hunk 5]
```diff
@@ -7,16 +7,18 @@ import (
...
 }
 
 // CreatorWithoutRepoList provides the type and registries.Creator to add to the registries Registry.
 // Populating the internal repo list will be disabled.
-func CreatorWithoutRepoList() (string, func(integration *storage.ImageIntegration) (types.Registry, error)) {
-	return "artifactory", func(integration *storage.ImageIntegration) (types.Registry, error) {
-		return docker.NewDockerRegistry(integration, true)
-	}
+func CreatorWithoutRepoList() (string, types.Creator) {
+	return "artifactory",
```
Comment (stehessel) [inline pkg/registries/artifactory/artifactory.go:20]: I'll create types for these names with the telemetry PR.

**Review by dhaus67 — APPROVED** [review]

**Comment by stehessel** [comment]
I'm afraid the qa test fails are not unrelated, will dig into those.

**Comment by stehessel** [comment]
Based on the qa tests failures, I found that artifact/gcr do support oauth bearer tokens when using our docker api client for all endpoints. I therefore added a gcp transport that translates the oauth token to basic credentials similar to the AWS one in #9031.

File: pkg/registries/docker/config.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+// Returns `Config.Transport` if it is set. Otherwise returns a default transport.
+func (c *Config) GetTransport() registry.Transport {
+	if c.Transport != nil {
+		return c.Transport
+	}
+	return DefaultTransport(c)
+}
+
+// DefaultTransport returns the default transport based on the configuration.
+func DefaultTransport(cfg *Config) registry.Transport {
```
Comment (RTann) [inline pkg/registries/docker/config.go:46]: want to wrap this all in a `once`?

File: pkg/registries/google/google.go [hunk 7]
```diff
@@ -32,50 +34,80 @@ func (g *googleRegistry) Match(image *storage.ImageName) bool {
...
 // Creator provides the type and registries.Creator to add to the registries Registry.
-func Creator() (string, func(integration *storage.ImageIntegration) (types.Registry, error)) {
-	return "google", func(integration *storage.ImageIntegration) (types.Registry, error) {
-		return NewRegistry(integration, false)
-	}
+func Creator() (string, types.Creator) {
+	return "google",
+		func(integration *storage.ImageIntegration, options ...types.CreatorOption) (types.Registry, error) {
+			cfg := types.ApplyCreatorOptions(options...)
+			return NewRegistry(integration, false, cfg.GetGCPTokenManager())
```
Comment (RTann) [inline pkg/registries/google/google.go:41]: so this is for all Google services? not just artifact registry?

**Review by RTann — COMMENTED** [review]
took an initial look at the image integration-specific parts. I'm hoping you can help clarify this:

so we no longer use the JSON way of access google registries? Because I see code in `cloudproviders`, it makes me think the wif stuff is for GCP users. What if someone isn't using GCP? Then can they no longer use google registries?

**Comment by RTann** [comment]
what does `WIF` stand for?

**Comment by stehessel** [comment]
Hi @RTann, let me try to give some background. WIF stands for workload identity federation. It is the Google framework that allows exchanging Kubernetes service account tokens for short-lived GCP cloud tokens. However, AWS and Azure have similar frameworks which are also called workload identities, STS tokens (secure token service) or a combination thereof.

The way this feature works in practice is that you have to setup the k8s service account -> cloud provider service account mapping. Once that is done, a pod can impersonate the workload identity with their K8s service account token + some configuration. For AWS and Azure the config is passed as env variables, for Google it is a service account json configured as a secret. That's why there is a manager for Google - it handles the secret config. In the Central integration, "use workload identity" is available as an authentication that replaces the static credentials. See for example the ECR integration, where this is already implemented and called "container IAM role".

Coming back to this PR, the following happens:
* The GCP token manager already exists and handles workload identity configuration for other integrations. The goal is to make it work with Artifact and GCR registries to enable a "use workload identity" button which replaces the need for static credentials.
* The token manager handles the token exchange and outputs a TokenSource with GCP access tokens.
* Note that pkg/cloudproviders/gcp/utils/util.go does the same even if workload identities are not enabled - then the access tokens are exchanged based on the static service account credentials. This is not technically needed - we could pass the whole service account json like before. But I think passing short-lived access tokens is preferable for security reasons.
* Other than this detail, this feature is meant to be backwards compatible. Meaning static credentials via the service account json are still supported, and using workload identities is an orthogonal, optional authentication method.

File: pkg/registries/google/google.go (hunk 7 above)
Comment (stehessel) [reply pkg/registries/google/google.go:41]: We want to support workload identities for all Google, Azure and AWS services. This PR deals with support for Artifact registry + Google container registry.

File: pkg/registries/docker/config.go (hunk 6 above)
Comment (stehessel) [reply pkg/registries/docker/config.go:46]: We call this method more than once - namely every time the access token has been rotated and we need to create a new transport with updated basic auth credentials.

File: pkg/registries/google/transport.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/stackrox/rox/pkg/registries/docker"
+	"github.com/stackrox/rox/pkg/sync"
+	"golang.org/x/oauth2"
+)
+
+var log = logging.LoggerForModule()
+
+// googleTransport represents a transport that converts an oauth token source
+// into docker registry credentials.
+// This is required kind of trickery is required because the docker API does not
```
Comment (dhaus67) [inline pkg/registries/google/transport.go:18]:
```suggestion
// This kind of trickery is required because the docker API does not
```

File: pkg/registries/google/transport.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+func (t *googleTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	if !t.token.Valid() {
+		if err := t.refreshNoLock(); err != nil {
+			return nil, err
+		}
+	}
```
Comment (dhaus67) [inline pkg/registries/google/transport.go:43]: We can use `concurrency.WithLock` here _iff_ it's safe to do `t.Transport.RoundTrip()` outside of the lock - which looking at everything should be fine. This way we minimize the time we actually hold the lock.

File: pkg/scanners/google/google.go [hunk 10]
```diff
@@ -66,6 +64,9 @@ func validate(google *storage.GoogleConfig) error {
 	if google.GetProject() == "" {
 		errorList.AddString("ProjectID must be specified for Google Container Analysis")
 	}
+	if google.GetWifEnabled() {
+		errorList.AddString("Workload identities are not supported for Scanner integrations")
+	}
```
Comment (dhaus67) [inline pkg/scanners/google/google.go:69]: Bit confused where this is now coming from, seems unrelated to the rest of the changes - care to add a comment _why_?

Comment (stehessel) [reply pkg/scanners/google/google.go:69]: Scanner and registries share the same image registration proto. We discussed that we don't want to support WIF for GCP scanners because they'll be deprecated/removed soon. That's why I added this check.

File: pkg/registries/google/transport.go (hunk 9 above)
Comment (stehessel) [reply pkg/registries/google/transport.go:43]: see follow up #9117
